Allow CI to run concurrent builds in master

alpeb opened the pull request:
Fixes #3911

Refactors the `cloud_integration` test to run in separate GKE clusters
that are created and torn down on the fly.
It leverages a new "gcloud" github action that is also used to set up
gcloud in other build steps (`docker_deploy` and `chart_deploy`).

The action also generates unique names for those clusters, based on the
git commit SHA and `run_id`, a recently introduced variable that is
unique per CI run and available to all the jobs.
This fixes part of #3635 in that CI runs on the same SHA don't interfere
with one another (in the `cloud_integration` test; still to do for
`kind_integration`).

The "gcloud" GH action is supported by `.github/actions/gcloud/index.js`
that has a couple of dependencies. To avoid having to commit
`node_modules`, after every change to that file one must run
```bash
# only needed the first time
npm i -g @zeit/ncc

cd .github/actions/gcloud
ncc build index.js
```
which generates the self-contained file
`.github/actions/gcloud/dist/index.js`.
(This last part might get easier in the future after other refactorings
outside this PR).

alpeb commented:
I've just realized we might be able to use a `post` hook in the `cloud_integration` test to tear down the cluster and thus get rid of the `cloud_cleanup` job. That hook is undocumented but you can see it in action for example with `actions/checkout`...

grampelberg commented on .github/actions/gcloud/index.js:
```diff
@@ -0,0 +1,81 @@
+      if (core.getInput('create')) {
+        await exec.exec('gcloud container clusters create',
+          [name, '--cluster-version', '1.15.7-gke.23', '--num-nodes', 1, '--machine-type',
+          'n1-standard-2', '--enable-network-policy']);
```
Can this be moved out into configuration? Feels like we'll want to change these without modifying some strings in the source. I'd recommend at least doing:

- `--preemptible`
- `--release-channel rapid`
- `--no-enable-stackdriver-kubernetes`
- `--no-enable-basic-auth`
- `--no-enable-legacy-authorization`

Should we open a separate ticket to have this test `--ha` as well?

alpeb replied:
Sounds good, I'll add those options. Yeah, I think `--ha` grants a separate ticket.

grampelberg commented on .github/actions/gcloud/index.js:
```diff
@@ -0,0 +1,81 @@
+const core = require('@actions/core');
```
As this appears to just be all `exec`. Should we just do it as a bash action instead? The added indirection feels weird.

alpeb replied:
I agree bash would be the right language here. But custom actions can only be built in javascript, which run directly on the Github VM, or a container that would run its own sandbox. The latter is geared more towards isolated tasks, whereas javascript actions modify the VM environment and can interact better with the other steps in the job.

I tried doing this with a bash script inside a container, that would set the gcloud, kubectl and docker configs in a directory from the VM that is mounted in the container, and then in the other job steps reconfigure those tools to point to those non-standard config locations, but it was messy and let to lots of repetition.

grampelberg replied:
Good call, I got the wrappers all confused.

kleimkuhler commented on .github/actions/gcloud/action.yml:
```diff
@@ -0,0 +1,21 @@
+    description: Set to true to create a new cluster. The linkerd CLI should have been installed in the runner beforehand.
+  destroy:
+    required: false
+    description: Set to false to create a new cluster. The linkerd CLI should have been installed in the runner beforehand.
```
Since this is not required, maybe this description should be `Set to true to destroy an existing cluster`?

alpeb replied:
Yea, that description was off. I'm gonna end up merging that into an `action` field, as per Siggy's suggestion below.

kleimkuhler replied:
Yep sounds good I like that plan!

kleimkuhler commented on .github/actions/gcloud/index.js:
```diff
@@ -0,0 +1,81 @@
+
+  // validate CLI version matches the repo
+  if (tag != clientVersion) {
+      throw `tag ${tag} differs from clientversion ${clientVersion}`
```
nit: `.. differs from clientVersion ..` or `.. differs from client version ..`?

kleimkuhler commented on .github/actions/gcloud/index.js:
```diff
@@ -0,0 +1,81 @@
+  // Last part is to distinguish runs on the same sha (run-id is unique per CI run).
+  // run-id has to be provided as an input because it turns out it's not available
+  // through github.context.run_id
+  const name = `testing-${tag}-${core.getInput('run-id')}`;
```
This is cool. Glad to see `run-id` being used here to solve the naming issue we have.

kleimkuhler commented on .github/workflows/workflow.yml:
```diff
@@ -471,34 +438,25 @@ jobs:
       if: github.ref == 'refs/heads/master' || startsWith(github.ref, 'refs/tags')
       uses: actions/checkout@v2
-    - name: Configure gcloud
+    - name: Install linkerd CLI
```
So it looks like `gcloud .. create/delete` only need the Linkerd CLI to be installed for the action's `getClusterName` which relies on `linkerd version --client --short`.

In `cloud_integration` this makes sense, but it's not stricly necessary to have the CLI installed for `cloud_cleanup`. Maybe instead of the descriptions requiring that a CLI is installed, they just pass in `$TAG` as an input? That way we don't need to install the CLI since we can already rely on `$TAG` having the correct value.

siggy replied:
Yeah agree, I think decoupling the gcloud github action from linkerd cli makes sense. It's good to validate the Linkerd CLI version matches repo version, but that could/should be done elsewhere. Taking `$TAG` as input a step further, maybe consider just having a `name` input to specify the cluster name.

siggy commented on .github/actions/gcloud/action.yml:
```diff
@@ -0,0 +1,21 @@
+    description: Set to true to create a new cluster. The linkerd CLI should have been installed in the runner beforehand.
+  destroy:
+    required: false
+    description: Set to false to create a new cluster. The linkerd CLI should have been installed in the runner beforehand.
```
should `create`/`delete` just be a single string input called `action`? feels like the api should not allow the user to set both of these bools to true or both to false. also consider documenting that if these are not set, the action's default behavior is just to configure the gcloud tooling.

alpeb replied:
Ok, an enum for this makes more sense indeed :ok_hand:

siggy commented on .github/workflows/workflow.yml:
```diff
@@ -361,28 +361,11 @@ jobs:
-        )
-        . "$dir/path.bash.inc"
-        gcloud auth configure-docker
-        bin/kubectl version --short
```
i think the removal of this code implies we can delete `bin/_gcp.sh`?

alpeb replied:
indeed :eye:

siggy commented on .github/actions/gcloud/action.yml:
```diff
@@ -0,0 +1,21 @@
+    description: Set to false to create a new cluster. The linkerd CLI should have been installed in the runner beforehand.
+  run-id:
+    required: false
+    default: ${{ github.run_id }}
```
💥

alpeb commented on .github/actions/gcloud/action.yml:
```diff
@@ -0,0 +1,21 @@
+name: gcloud
```
Definitely. Itself could also go through CI to avoid having to do the `ncc` step manually.

alpeb commented on .github/actions/gcloud/index.js:
```diff
@@ -0,0 +1,81 @@
+  });
+
+  // validate CLI version matches the repo
+  if (tag != clientVersion) {
```
Good catch :+1: 
`eslint` doesn't make part yet of CI. But as per the previous comment, I can add that step into the workflow that will publish this action as standalone :wink:

alpeb commented on .github/actions/gcloud/index.js:
```diff
@@ -0,0 +1,81 @@
+      }
+  });
+
+  await exec.exec(`${process.env.HOME}/.linkerd version --client --short`, [], {
```
I'll be getting rid of `getClusterName()` in my next push, but still nevertheless a good point :+1:

alpeb commented:
Thanks @grampelberg, @kleimkuhler and @siggy for the detailed review!
I've just pushed another iteration with the following changes:
- Got rid of cloud_cleanup by using a post hook in the gcloud action
- Removed cluster naming responsibility from the gcloud action
- Added more options to the gcloud action to better control the type of cluster created, with some sane defaults
- Removed `bin/_gcp.sh`
- Consolidated `.gitignore` statements

Once we're all ok with these changes, I'll proceed to move `actions/gcloud` into its own repo with a little CI for linting and for generating `dist/index.js` automatically.

alpeb commented:
Note: GH is still waiting for the "Cloud cleanup" step, but that's gone. Probably the same issue #4002 was having, i.e.  "Branch Protection Rules" that need to be modified by a user with enough access.

kleimkuhler commented on .github/actions/gcloud/action.yml:
```diff
@@ -0,0 +1,47 @@
+  using: node12
+  main: dist/index.js
+  post: dist/index.js
+  post-if: always()
```
Does this `post_if` condition ensure the action runs again and hits the `isPost` branch that deletes the cluster?

alpeb replied:
yep, that's correct. I did test having the cluster creation job fail, and the deletion post step does get triggered.

siggy commented on .github/actions/gcloud/index.js:
```diff
@@ -0,0 +1,110 @@
+      } else {
+        await destroy();
+      }
+    }
```
can we avoid having this `isPost` state, and instead just call separate endpoints for `runs/main` and `runs/post`?

alpeb replied:
I followed the pattern used by `actions/checkout`, but you're right, that's unneeded complexity. I'll see if I can split this up.

siggy commented on .github/actions/gcloud/action.yml:
```diff
@@ -0,0 +1,47 @@
+    default: false
+  enable_legacy_auth:
+    required: false
+    default: false
```
are all these extra flags intended to make this a drop-in replacement for another package? if not, consider dropping.

alpeb replied:
Like a replacement for an existing 3rd party gcloud GH action? Nope, just following up on a [request](https://github.com/linkerd/linkerd2/pull/4001#discussion_r373560269) from @grampelberg

alpeb commented:
Ok in my last few pushes I've spun-off the "destroy" post-action and moved all the action into a separate repo https://github.com/linkerd/linkerd2-action-gcloud, that itself has the CI to build the consolidated js files and full linting.
Please give it a final look before I merge :pray:

alpeb commented:
> This all looks great !  
> 
> Have you been able to test most of these jobs? Possibly by disabling some `if: startsWith(github.ref, 'refs/tags')` checks?

Yeah, I've been actually testing in a fork, using PRs and merging into the fork's master, using the appropriate secrets :+1:
